### PR TITLE
ifdef out unsupported Enum underlying types for nativeaot

### PR DESCRIPTION
**Pull request by stephentoub**
@jkotas, worthwhile?

Contributes to https://github.com/dotnet/runtime/issues/79437

### CODE REVIEW

**Comment by jkotas** [comment]
It looks fine to me. I would make a new #define for it at the top of the file to make it more self-documenting. We may want to enable the define on other targets where we care a lot about size a lot, like wasm. Thank you for looking for ways to fix the size regression!

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs [hunk 1]
```diff
@@ -1,6 +1,11 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+// Rare enums types not expressible in C# are not supported in native AOT
+#if !NATIVEAOT
+#define RARE_ENUMS
```
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: @SamMonoRT @kotlarmilos should this also include all mobile/web targets?

**Comment by jkotas** [comment]
@MichalStrehovsky PTLA

This fixes some of the size regression by ifdefing out support for rare enums, to match what we used to do in native AOT before the recent Enum changes.

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs (hunk 1 above)
Comment (SamMonoRT) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: Yes, if we can include the mobile/web targets to mitigate the app size regressions, we should include those now, till we are able to identify and fix the Mono AOT compiler limitations.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: I have pushed a commit to disable the rare enums support for Mono as well. Please take a look.

**Review by MichalStrehovsky — APPROVED** [review]
Looks good, although I've always been puzzled by our position on these. We care, but also we don't care.

I'd honestly just delete the reflection support for these and keep things simple, because it's either important and it was a bug that CoreCLR-AOT didn't have this, or it's not important and people won't miss it on CoreCLR-JIT either. (Unless there's something unique for CoreCLR-JIT, like C++/CLI support, that makes it needed there.)

Do we need a breaking change label for Mono?

**Comment by jkotas** [comment]
> unique for CoreCLR-JIT, like C++/CLI support

Yes, C++/CLI support, obfuscated code and compat for niche languages.

e.g. this compiles fine in C++/CLI:
```
enum MyEnum : intptr_t
{
	Red, Green, Blue
};
```

**Comment by jkotas** [comment]
> `enum MyEnum : intptr_t { Red, Green, Blue };

This example is fine actually - C/C++ will lower the underlying type to int32/int64. A better example:
```
enum MyEnum : bool
{
	Red, Blue
};
```

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs (hunk 1 above)
Comment (kotlarmilos) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: I've inspected native symbols of an iOS app and non-integral types except `bool` are supported in the Mono AOT engine.

**Comment by SamMonoRT** [comment]
> > Do we need a breaking change label for Mono?
> 
> @SamMonoRT How would like to handle that?

Adding @lambdageek and @vargaz for their opinion.

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: Ok, I am going to keep the rare enums enabled for Mono. I do not data to make the call on whether to disable it for Mono or not. It is two line change to enable/disable this for given config.

Comment (SamMonoRT) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: Sounds good, leave it enabled for now. We can run some measurements and see if it helps in size if we disable it.

Comment (kotlarmilos) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:6]: It doesn't look to me like breaking change - by doing that we will restrict some theoretical or borderline cases. I would suggest excluding them as they are not expressible in C# and such change could mitigate the size regression. Some edge cases might fail where non-integral enums are created by using reflection, which is not recommended by the documentation. 

Let's hear the other opinions as well. If it is a blocker, we can proceed and update it once we align.

**Comment by jkotas** [comment]
All known errors according to the build analysis.

**Pull request merged**